l2geth/rollup/sync_service.go

```diff
@@ -630,100 +509,187 @@ func (s *SyncService) SetLatestVerifiedIndex(index *uint64) {
 	if block == nil {
-		err := s.applyTransaction(tx)
-		if err != nil {
-			return fmt.Errorf("Maybe apply transaction failed on index %d: %w", *index, err)
-		}
-		return nil
+		return fmt.Errorf("Block %d is not found", *index+1)
 	}
-	// There is already a transaction at that index, so check
-	// for its equality.
 	txs := block.Transactions()
 	if len(txs) != 1 {
-		log.Info("block", "txs", len(txs), "number", block.Number().Uint64())
-		return fmt.Errorf("More than 1 transaction in block")
+		return fmt.Errorf("More than one transaction found in block %d", *index+1)
 	}
-	if isCtcTxEqual(tx, txs[0]) {
-		log.Info("Matching transaction found", "index", *index)
+	if !isCtcTxEqual(tx, txs[0]) {
+		log.Error("Mismatched transaction", "index", *index)
 	} else {
-		log.Warn("Non matching transaction found", "index", *index)
+		log.Debug("Historical transaction matches", "index", *index, "hash", tx.Hash().Hex())
 	}
 	return nil
 }
 
-// Lower level API used to apply a transaction, must only be used with
-// transactions that came from L1.
-func (s *SyncService) applyTransaction(tx *types.Transaction) error {
+// applyTransactionToTip will do sanity checks on the transaction before
+// applying it to the tip. It blocks until the transaction has been included in
+// the chain. It is assumed that validation around the index has already
+// happened.
+func (s *SyncService) applyTransactionToTip(tx *types.Transaction) error {
+	if tx == nil {
+		return errors.New("nil transaction passed to applyTransactionToTip")
+	}
+	// Queue Origin L1 to L2 transactions must have a timestamp that is set by
+	// the L1 block that holds the transaction. This should never happen but is
+	// a sanity check to prevent fraudulent execution.
+	if tx.QueueOrigin().Uint64() == uint64(types.QueueOriginL1ToL2) {
+		if tx.L1Timestamp() == 0 {
+			return fmt.Errorf("Queue origin L1 to L2 transaction without a timestamp: %s", tx.Hash().Hex())
+		}
+	}
+	// If there is no OVM timestamp assigned to the transaction, then assign a
+	// timestamp and blocknumber to it. This should only be the case for queue
+	// origin sequencer transactions that come in via RPC. The L1 to L2
+	// transactions that come in via `enqueue` should have a timestamp set based
+	// on the L1 block that it was included in.
+	// Note that Ethereum Layer one consensus rules dictate that the timestamp
+	// must be strictly increasing between blocks, so no need to check both the
+	// timestamp and the blocknumber.
+	if tx.L1Timestamp() == 0 {
+		ts := s.GetLatestL1Timestamp()
+		bn := s.GetLatestL1BlockNumber()
+		tx.SetL1Timestamp(ts)
+		tx.SetL1BlockNumber(bn)
+	} else if tx.L1Timestamp() > s.GetLatestL1Timestamp() {
+		// If the timestamp of the transaction is greater than the sync
+		// service's locally maintained timestamp, update the timestamp and
+		// blocknumber to equal that of the transaction's. This should happen
+		// with `enqueue` transactions.
```

Comment:
You say that this should happen with `enqueue` transactions. Should we add a sanity check that the `tx.QueueOrigin().Uint64() == uint64(types.QueueOriginL1ToL2)` in this case then? Or is it guaranteed that `tx.L1Timestamp() != 0` only in the case that the transaction is an `enqueue`?

Reply:
It is guaranteed that `tx.L1Timestamp() == 0` only for `types.QueueOriginSequencer` transactions, assuming no bugs in the upstream software